nums/core/array/blockarray.py

```diff
@@ -883,21 +885,41 @@ def _block_shape_reshape(self, arr, block_shape):
     def _strip_ones(self, shape):
         return tuple(filter(lambda x: x != 1, shape))
 
-    def _is_simple_reshape(self, arr: BlockArray, shape, block_shape):
+    def _is_simple_reshape(self, arr: BlockArray, shape):
         # Is the reshape a difference of factors of 1?
         # Strip out 1s and compare.
-        return self._strip_ones(shape) == self._strip_ones(
-            arr.shape
-        ) and self._strip_ones(block_shape) == self._strip_ones(arr.block_shape)
+        return self._strip_ones(shape) == self._strip_ones(arr.shape)
 
     def _simple_reshape(self, arr, shape, block_shape):
         # Reshape the array of blocks only.
         # This is only used when the difference in shape are factors of 1s,
         # and the ordering of other factors are maintained.
 
-        # Check assumptions.
-        assert len(self._strip_ones(arr.shape)) == len(self._strip_ones(shape))
+        # Strip ones
+        arr_shape_stripped = self._strip_ones(arr.shape)
+        new_shape_stripped = self._strip_ones(shape)
 
+        # Check assumptions.
+        assert len(arr_shape_stripped) == len(new_shape_stripped)
+
+        if block_shape == shape:
```

Comment:
All tests / validations of this reshape procedure from line 905 to 922 should be moved to `_is_simple_reshape`; it should test whether the array should be reshaped using this procedure. If the array cannot be reshaped with this procedure and we're already in this procedure, then we have to raise instead of falling back to a more general reshape routine.